platforms/raspi/raspi_adaptor.go

```diff
@@ -100,18 +116,31 @@ func (r *Adaptor) DigitalPin(pin string, dir string) (sysfsPin sysfs.DigitalPinn
 		return
 	}
 
-	if r.digitalPins[i] == nil {
-		r.digitalPins[i] = sysfs.NewDigitalPin(i)
-		if err = r.digitalPins[i].Export(); err != nil {
-			return
-		}
+	currentPin, err := r.getExportedDigitalPin(i, dir)
+
+	if err != nil {
+		return
 	}
 
-	if err = r.digitalPins[i].Direction(dir); err != nil {
+	if err = currentPin.Direction(dir); err != nil {
 		return
 	}
 
-	return r.digitalPins[i], nil
+	return currentPin, nil
+}
+
+func (r *Adaptor) getExportedDigitalPin(translatedPin int, dir string) (sysfsPin sysfs.DigitalPinner, err error) {
```

Comment:
I extracted this part to minimize the mutex locking and make it easier to implement.
Doing so in `DigitalPin` would require either to lock `currentPin.Direction` or unlock the mutex without defer with a more error prone code.
This way `DigitalPin` just call `getExportedDigitalPin` without using the mutex in an explicit way.

If you prefer I can inline the new function, but I think this way is better